File: server/routes/sessions.py
import logging
from fastapi import APIRouter
from Coder.server.schemas import SessionCreate
from Coder.tools.session_manager import SessionManager
from Coder.tools.file_saver import FileSaver

logger = logging.getLogger(__name__)
router = APIRouter()
mgr = SessionManager()


@router.get("/")
async def list_sessions():
    sessions = mgr.list_sessions()
    return {"sessions": sessions}


@router.post("/")
async def create_session(body: SessionCreate = None):
    session = mgr.create_session(title=body.title if body else None)
    return session


@router.get("/{session_id}")
async def get_session(session_id: str):
    session = mgr.get_session(session_id)
    if not session:
        return {"error": "Session not found"}
    return session


@router.get("/{session_id}/messages")
async def get_messages(session_id: str):
    checkpointer = FileSaver()
    messages = mgr.get_session_messages_from_checkpoint(session_id, checkpointer)
    return {"messages": messages}


@router.delete("/{session_id}")
async def delete_session(session_id: str):
    ok = mgr.delete_session(session_id)
    return {"deleted": ok}

File: Coder/web/src/api/sessions.ts
import { api } from './client'
import type { Session, Message } from '../types'

export async function listSessions(courseId?: string): Promise<Session[]> {
  const query = courseId ? `?course_id=${encodeURIComponent(courseId)}` : ''
  const data = await api.get<{ sessions: Session[] }>(`/sessions/${query}`)
  return data.sessions
}

export async function createSession(title?: string, courseId?: string): Promise<Session> {
  const body: Record<string, string> = {}
  if (title) body.title = title
  if (courseId) body.course_id = courseId
  return api.post<Session>('/sessions/', body)
}

export async function getMessages(sessionId: string): Promise<Message[]> {
  const data = await api.get<{ messages: Message[] }>(`/sessions/${sessionId}/messages`)
  return data.messages
}

export async function deleteSession(sessionId: string) {
  return api.del(`/sessions/${sessionId}`)
}
